### DISCUSSION TITLE
Is there a way to provide a custom SQL query for an entity, similar to Entity Framework Core's ToSqlQuery method?

### DISCUSSION THREAD

**Discussion by yankarinRG** [linq2db/linq2db#4692] [Q&A]
Hello,
In Entity Framework Core, there's a method called [ToSqlQuery](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationalentitytypebuilderextensions.tosqlquery?view=efcore-8.0#microsoft-entityframeworkcore-relationalentitytypebuilderextensions-tosqlquery(microsoft-entityframeworkcore-metadata-builders-entitytypebuilder-system-string)) that allows configuring a custom SQL query to be the source for an entity type. Here's an example of how it works:
```
public class Blog
{
    public int Id { get; set; }
    public string Title { get; set; }
}

protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<Blog>(builder =>
    {
        builder.ToSqlQuery("SELECT B.Id, BT.Title FROM Blog B LEFT JOIN BlogTitle BT ON BT.BlogId = B.Id");
    });
}
```
With this configuration, the entity type `Blog` will be populated based on the results of the provided custom SQL query.

For example:
`var blogs = context.Set<Blog>().ToList();`

This would execute the custom SQL query (`SELECT B.Id, BT.Title FROM Blog B LEFT JOIN BlogTitle BT ON BT.BlogId = B.Id`) and return the results mapped to the `Blog` entity.

Question: Is there a similar API or method available in LINQ To DB that allows us to configure a custom SQL query as the data source for an entity type like in the `ToSqlQuery` method from EF Core? If not, how would I go about solving this use case in LINQ To DB where I need to map the results of a custom SQL query to a specific entity?

Thank you in advance!

**Comment by viceroypenguin** [comment]
Yes, you can use `context.FromSql<Blog>("SELECT B.Id, BT.Title FROM Blog B LEFT JOIN BlogTitle BT ON BT.BlogId = B.Id")`, which will give you an `IQueryable<Blog>` that you can use as a base for a larger LINQ query.

```cs
public sealed class DbContext
{
  public IQueryable<Blog> Blog => this..FromSql<Blog>("SELECT B.Id, BT.Title FROM Blog B LEFT JOIN BlogTitle BT ON BT.BlogId = B.Id");
}
```

**Comment by yankarinRG** [comment]
Thank you for the clarification!

The `FromSql` API is indeed similar to Entity Framework Core's `FromSql` method (e.g., `context.Set<Blog>().FromSql("SELECT B.Id, BT.Title FROM Blog B LEFT JOIN BlogTitle BT ON BT.BlogId = B.Id")`), which is great for building LINQ queries on top of custom SQL. However, what I'm looking for is more akin to EF Core’s `ToSqlQuery` method.

My exact use case involves migrating away from an in-house ORM that uses attribute-based mapping, where the defining query for an entity is specified via an attribute on the class (similar to the table mapping). For example:
```
[Query("SELECT B.Id, BT.Title FROM Blog B LEFT JOIN BlogTitle BT ON BT.BlogId = B.Id")]
public class Blog
{
    public int Id { get; set; }
    public string Title { get; set; }
}
```

So what I was hoping to find in LINQ To DB is something closer to EF Core’s `ToSqlQuery`, where the SQL query is directly tied to the entity's source definition, to avoid manually handling entities with the Query attribute in the custom ORM implementation, like this:
```
private static readonly ConcurrentDictionary<Type, string> _queryCache = new();

public List<T> GetAll<T>() where T : class
{
    return GetAllImpl<T>().ToList();
}

private IEnumerable<T> GetAllImpl<T>() where T : class
{
    if (_queryCache.GetOrAdd(typeof(T), GetQueryNoCache) is { } query)
    {
        return _dataConnection.Query<T>(query);
    }

    return _dataConnection.GetTable<T>();
}

private static string GetQueryNoCache(Type type)
{
    return type.GetCustomAttribute<QueryAttribute>()?.Sql;
}
```

Currently, I'm using a custom `IMetadataReader` to read our attributes and map them to LINQ To DB (though I'm not entirely sure if this is the optimal approach). I was wondering if there's a built-in attribute or mechanism to define the source query directly on the entity class, which could simplify the process.

If there isn't an out-of-the-box solution, I can work with the current workaround, though it feels a bit hack-ish. That said, I wanted to check if there’s a more streamlined or official approach available before settling on it.

Thanks again for your help!

**Reply by viceroypenguin** [reply]
There is no built-in equivalent in the configuration that uses a sql as part of the entity definition. That's why I proposed using `FromSql` as a property on the context, so that consumers will get the correct definition that way instead of seeing a table directly.

**Reply by yankarinRG** [reply]
Thank you for the suggestion. Unfortunately, in the custom ORM we're using, there are no compile-time tables like `ITable` in LINQ To DB or `DbSet` in EF Core. All entity retrieval happens dynamically through methods like `GetAll<T>()` or `Get<T>(params object[] keyValues)`, as shown in the snippet above. Because of this, I can't use the approach of exposing `ITable` for each entity (DB table). 

This setup is something I've inherited, and unfortunately, it's not possible to change the existing interface contract to adopt a more traditional approach with `ITable` or similar constructs. Given this constraint, looks like I’ll just go with the above idea—determine if it’s a Query entity or a standard entity, perform the data retrieval, and call it good.

Thanks again for your help!

**Reply by viceroypenguin** [reply]
You bet! Good luck. Sorry you can't use `async` or `IQueryable<>`

**Comment by igor-tkachev** [comment]
```c#
using System;

using LinqToDB;
using LinqToDB.Data;
using LinqToDB.DataProvider.SQLite;

namespace GenericLinqToDB
{
	class Program
	{
		static void Main(string[] args)
		{
			DataConnection.TurnTraceSwitchOn();
			DataConnection.WriteTraceLine = (s, s1, _) => Console.WriteLine(s, s1);

			var options = new DataOptions()
				.UseSQLite("Data Source=TestData.sqlite", SQLiteProvider.Microsoft);

			using var db = new DataConnection(options);

			db.CreateTable<Blog>     (tableOptions:TableOptions.CheckExistence);
			db.CreateTable<BlogTitle>(tableOptions:TableOptions.CheckExistence);

			_ = GetAll<BlogQuery>(db);
		}

		public class Blog
		{
			public int Id { get; set; }
		}

		public class BlogTitle
		{
			public int    Id    { get; set; }
			public string Title { get; set; }
		}

		interface IQueryableEntity<T>
		{
			static abstract IQueryable<T> Query(DataConnection db);
		}

		class BlogQuery : IQueryableEntity<BlogQuery>
		{
			public int    Id    { get; set; }
			public string Title { get; set; }

			public static IQueryable<BlogQuery> Query(DataConnection db) =>
				from b in db.GetTable<Blog>()
				join BlogTitle in db.GetTable<BlogTitle>() on b.Id equals BlogTitle.Id
				select new BlogQuery
				{
					Id    = b.Id,
					Title = BlogTitle.Title
				};
		}

		static List<T> GetAll<T>(DataConnection db) where T : class, IQueryableEntity<T>
		{
			return T.Query(db).ToList();
		}
	}
}
```

**Reply by yankarinRG** [reply]
Thank you for the elegant solution! I really appreciate your time in putting that together. Unfortunately, there are a few limitations on my end that prevent me from adopting this approach:

1. I need to pick up the entity's data source definition from the `QueryAttribute` that's already in place, rather than defining the query in the `Query` method itself.
2. I'd need to reference LINQ To DB in the models project, but since LINQ To DB is considered an implementation detail in our architecture, I'm trying to avoid adding it there.
3. A minor issue (which could be addressed with the `LangVersion` property) is that we’re on .NET Framework, so we're limited to C# 7.3, while static abstract members in interfaces require C# 11.

That said, I’m actually trying to avoid the approach of manually handling entities decorated with `QueryAttribute` and maintaining two branches (one for standard entities and one for custom queries), as I showed earlier. I’m still digging deeper into LINQ To DB to see if there’s any specialized or hidden functionality that might help retrofit my use case. So far, I’ve come across attributes like `[Sql.TableExpression("OPENJSON({2}, {3}) {1}")]` and `[ExpressionMethod(nameof(GenerateOpenJsonStringImpl))]`, which look promising but require further experimentation.
